File: backend/leetcode/p1/python/twoSum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        pass

## Two Sum Brute Force Solution
def twoSumBrute();
    n = len(nums)
    for i in range(n - 1):
        for j in range(i + 1, n):
            if nums[i] + nums[j] == target:
                return [i, j]
    return []  # No solution found

## Two Sum Two Pointer Solution O(nlogn)
def twoSumTwoPointers():
    # Create a list of tuples (number, index)
    nums = [(num, i) for i, num in enumerate(nums)]
    # Sort the list of tuples by number
    nums.sort()
    # Initialize two pointers
    left, right = 0, len(nums) - 1
    while left < right:
        # Calculate the sum of the two numbers
        total = nums[left][0] + nums[right][0]
        if total == target:
            return [nums[left][1], nums[right][1]]
        elif total < target:
            left += 1
        else:
            right -= 1
    return []  # No solution found

## Two Sum HashMap Solution
def twoSumHashMap():
    # Create a dictionary to store the index of each number
    num_to_index = {}
    for i, num in enumerate(nums):
        if target - num in num_to_index:
            return [num_to_index[target - num], i]
        num_to_index[num] = i
    return []  # No solution found

File: backend/leetcode/p1/java/TwoSum.java
import java.util.Arrays;

public class TwoSum {
    public static void main(String[] args) {
        System.out.println(TestCaseOne());
        System.out.println(TestCaseTwo());
        System.out.println(TestCaseThree());
    }

    private static boolean TestCaseOne() {
        int[] nums = new int[]{2,7,11,15};
        int target = 9;
        int[] solution = new int[]{0, 1};

        return GeneralTestCase(nums, target, solution);
    }

    private static boolean TestCaseTwo() {
        int[] nums = new int[]{3,2,4};
        int target = 6;
        int[] solution = new int[]{1, 2};

        return GeneralTestCase(nums, target, solution);
    }

    private static boolean TestCaseThree() {
        int[] nums = new int[]{3, 3};
        int target = 6;
        int[] solution = new int[]{0, 1};

        return GeneralTestCase(nums, target, solution);
    }

    private static boolean GeneralTestCase(int[] nums, int target, int[] solution) {
        TwoSumBruteForce twoSumBruteForce = new TwoSumBruteForce();
        TwoSumTwoPointers twoSumTwoPointers = new TwoSumTwoPointers();
        TwoSumHashMap twoSumHashMap = new TwoSumHashMap();

        boolean bruteForceWorks = Arrays.equals(twoSumBruteForce.twoSum(nums, target), solution);
        boolean twoPointerWorks = Arrays.equals(twoSumTwoPointers.twoSum(nums, target), solution);
        boolean twoSumHashMapWorks = Arrays.equals(twoSumHashMap.twoSum(nums, target), solution);

        System.out.println("bruteForceWorks: " + bruteForceWorks);
        System.out.println("twoPointerWorks: " + twoPointerWorks);
        System.out.println("twoSumHashMapWorks: " + twoSumHashMapWorks);

        return bruteForceWorks && twoPointerWorks && twoSumHashMapWorks;
    }
}
